### PR TITLE
Generating copy-on-rewrite logic

### PR DESCRIPTION
**Pull request by systay**
## Description
Adds a new syntax tree rewrite utility. It makes it possible to share parts of the syntax tree with the original, so it should decrease memory pressure somewhat during planning. It also traverses the tree only once to do both operations, before we first needed to clone the full tree and only after could we rewrite it.

Additionally, it can report back to the user which nodes got cloned, which makes it possible to copy semantic information between the two trees.

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/sqlparser/cow.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	cow struct {
+		pre    func(node, parent SQLNode) bool
+		post   func(cursor *CopyOnWriteCursor)
+		cloned func(old, new SQLNode)
+		cursor CopyOnWriteCursor
+	}
+)
+
+func CopyOnRewrite(node SQLNode, pre func(node, parent SQLNode) bool, post func(cursor *CopyOnWriteCursor)) SQLNode {
```
Comment (systay) [inline go/vt/sqlparser/cow.go:34]: comment

File: go/vt/sqlparser/cow.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	c.cursor.replaced = nil
+	c.post(&c.cursor)
+	if c.cursor.replaced != nil {
+		return c.cursor.replaced, true
+	}
+	return node, false
+}
+
+// StopTreeWalk aborts the current tree walking. No more nodes will be visited, and the rewriter will exit out early
+func (c *CopyOnWriteCursor) StopTreeWalk() {
```
Comment (systay) [inline go/vt/sqlparser/cow.go:76]: Maybe we should use a recovered panic here instead of setting a boolean and checking it everywhere

**Review by vmg — APPROVED** [review]
Looks great. I discussed with @systay the unified interface for `pre`/`post` and I think this is a great design!

File: go/vt/sqlparser/cow.go (hunk 2 above)
Comment (frouioui) [reply go/vt/sqlparser/cow.go:57]: I thought that `recover` was a source of performance loss

**Review by frouioui — APPROVED** [review]

**Pull request merged**